zqd/handlers.go

```diff
@@ -78,59 +77,25 @@ func handlePacketSearch(c *Core, w http.ResponseWriter, r *http.Request) {
 	}
 	defer cancel()
 
-	req := &api.PacketSearch{}
+	var req api.PacketSearch
 	if err := req.FromQuery(r.URL.Query()); err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
-	}
-	if s.PacketPath() == "" || !s.HasFile(ingest.PcapIndexFile) {
-		http.Error(w, "space has no pcaps", http.StatusNotFound)
-		return
-	}
-	index, err := pcap.LoadIndex(s.DataPath(ingest.PcapIndexFile))
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	f, err := os.Open(s.PacketPath())
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
-	defer f.Close()
-	slicer, err := pcap.NewSlicer(f, index, req.Span)
+	reader, closer, err := s.PcapSearch(req)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
-	pcapReader, err := pcapio.NewReader(slicer)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-
-	var search *pcap.Search
-	switch req.Proto {
-	default:
-		msg := fmt.Sprintf("unsupported proto type: %s", req.Proto)
-		http.Error(w, msg, http.StatusBadRequest)
-		return
-	case "tcp":
-		flow := pcap.NewFlow(req.SrcHost, int(req.SrcPort), req.DstHost, int(req.DstPort))
-		search = pcap.NewTCPSearch(req.Span, flow)
-	case "udp":
-		flow := pcap.NewFlow(req.SrcHost, int(req.SrcPort), req.DstHost, int(req.DstPort))
-		search = pcap.NewUDPSearch(req.Span, flow)
-	case "icmp":
-		search = pcap.NewICMPSearch(req.Span, req.SrcHost, req.DstHost)
-	}
+	defer closer.Close()
 	w.Header().Set("Content-Type", "application/vnd.tcpdump.pcap")
-	w.Header().Set("Content-Disposition", fmt.Sprintf("inline; filename=%s.pcap", search.ID()))
-	err = search.Run(ctx, w, pcapReader)
+	w.Header().Set("Content-Disposition", fmt.Sprintf("inline; filename=%s.pcap", reader.ID()))
+	_, err = reader.WriteToWithContext(ctx, w)
 	if err != nil {
 		if err == pcap.ErrNoPacketsFound {
 			http.Error(w, err.Error(), http.StatusNotFound)
 		} else {
-			http.Error(w, err.Error(), http.StatusBadRequest)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
```

Comment:
This call to `http.Error()` will be a no-op if the `reader.WriteToWithContext()` call above resulted in any `Write` to the http.ResponseWriter. 

There's a similar issue in the `/search` handler, for which I have a branch that I haven't yet turned into a PR (https://github.com/brimsec/zq/tree/fix-search-error).

The pattern I've used there is that the handler does request validation, returning an HTTP error upon failure. If validation is ok, then the handler calls out to the relevant function (in this case WriteToWithContext) and the HTTP request gets a 2xx back, and  subsequent api protocol messages. If tehre is an error during processing, it is carried in the final `TaskEnd` message.

Reply:
Yep that's a good point @henridf 

You can't really use `EndTask` here because the content-type has already been set to pcap. Hmm.

We also have a problem with the `pcap.ErrNoPacketsFound` because again the content headers have been set... I think. Also `pcap.Search` will still write pcap section headers even if no results are found. Ugh.

Let me suggest that I file these concerns as a separate ticket and follow up on a separate pr? I think we do need to address this.

Reply:
> You can't really use EndTask here because the content-type has already been set to pcap.

Oh right. Addressing this is more complicated than for the search endpoint... so yes please file a separate ticket and it can be done separately.

Reply:
Excellent insight, @henridf!

Reply:
Filed #550 and #551